etl/analytics.py

```diff
@@ -0,0 +1,833 @@
+    date_min: str = DATE_MIN,
+    date_max: str = DATE_MAX,
+) -> pd.DataFrame:
+    """
+    Fetch number of chart views (renders) per chart from Metabase, optionally filtered by chart IDs and date range.
+
+    Parameters
+    ----------
+    chart_ids : list of int, optional
+        List of chart IDs to filter the results. If None, all charts are included.
+    date_min : str, optional
+        Minimum date to filter the results. If None, no minimum date is applied.
+    date_max : str, optional
+        Maximum date to filter the results. If None, no maximum date is applied.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of chart views per chart.
+
+    """
+    where_clauses = []
+    if date_min:
+        where_clauses.append(f"v.day >= '{date_min}'")
+    if date_max:
+        where_clauses.append(f"v.day <= '{date_max}'")
+    if chart_ids:
+        id_list = ", ".join(str(cid) for cid in chart_ids)
+        where_clauses.append(f"c.chart_id IN ({id_list})")
+
+    where_sql = "WHERE " + " AND ".join(where_clauses) if where_clauses else ""
+
+    query = f"""
+    SELECT
+        c.chart_id,
+        c.url,
+        c.published_at,
+        SUM(v.events) AS views
+    FROM charts c
+    JOIN grapher_views_detailed v ON c.url = v.grapher
+    {where_sql}
+    GROUP BY c.chart_id, c.url, c.published_at
+    ORDER BY views DESC
+    """
+    df_views = read_metabase(sql=query)
+
+    # To calculate the average daily views, we need to figure out the number of days for which we are counting views.
+    # This will be either the first date since we have analytics (DATE_MIN) or the publication date of the chart (if more recent than that).
+    published_at = pd.to_datetime(df_views["published_at"])
+    start_date = published_at.where(published_at > pd.to_datetime(date_min), pd.to_datetime(date_min))
+
+    # Add a column with the number of days that the views are referring to.
+    df_views["n_days"] = (pd.to_datetime(date_max) - start_date).dt.days
+
+    # Add a column for the average number of daily views.
+    df_views["views_daily"] = df_views["views"] / df_views["n_days"]
+
+    # Fix infs (for charts that were published in the last day).
+    df_views.loc[df_views["views_daily"] == float("inf"), "views_daily"] = 0
+
+    return df_views
+
+
+def get_chart_views_last_n_days(
+    chart_ids: Optional[List[int]] = None,
+    n_days: int = 30,
+) -> pd.DataFrame:
+    """
+    Fetch number of chart views per chart for the last n_days.
+
+    Parameters
+    ----------
+    chart_ids : list of int, optional
+        List of chart IDs to filter the results. If None, all charts are included.
+    n_days : int
+        Number of days to look back for views. Default is 30 days.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of chart views per chart for the last n_days.
+
+    """
+    # Calculate date range.
+    date_max = str(datetime.today().date())
```

Comment:
same problem as previous comment that `date_max` won't be the same as the max date in `grapher_views_detailed`. So the retrieved days of data here will be less than `n_days`. 

Can fix with something like `date_max = read_metabase(sql="select max(day) from grapher_views_detailed")`